src/Microsoft.AspNetCore.SignalR.Common/Protocol/HandshakeProtocol.cs

```diff
@@ -17,28 +19,39 @@ public static class HandshakeProtocol
     {
         private const string ProtocolPropertyName = "protocol";
         private const string ProtocolVersionPropertyName = "version";
+        private const string MinorVersionName = "minorVersion";
         private const string ErrorPropertyName = "error";
         private const string TypePropertyName = "type";
 
-        /// <summary>
-        /// The serialized representation of a success handshake.
-        /// </summary>
-        public static ReadOnlyMemory<byte> SuccessHandshakeData;
+        [Obsolete()]
+        public static ReadOnlyMemory<byte> SuccessHandshakeData = new byte[] { 123, 34, 109, 105, 110, 111, 114, 86, 101, 114, 115, 105, 111, 110, 34, 58, 48, 125, 30 }; 
+
+
+        private static Dictionary<IHubProtocol, ReadOnlyMemory<byte>> _messageCache = new Dictionary<IHubProtocol, ReadOnlyMemory<byte>>();
 
-        static HandshakeProtocol()
+        public static ReadOnlySpan<byte> GetCachedSuccessMessage(IHubProtocol protocol)
         {
+            if (_messageCache.TryGetValue(protocol, out var message))
```

Comment:
A regular dictionary isn't threadsafe. If a message is added at the same time as one is fetched then an error can occur.

Reply:
Yeah, we should use a `ConcurrentDictionary` here, or only access the dictionary from a lock.

Reply:
yeah it seems this is breaking the CI tests.

Will change it to a `ConcurrentDictionary`.

Ideally, would cache messages on startup from a list of "all supported protocols", rather than having to check and maybe cache on sending a response. Not sure where to find an "all supported protocols" list though, it seems each hub has its own. Could try something like caching any new protocols on hub creation, but that seems like more trouble than it's worth.